src/number_theory/modular.lean

```diff
@@ -0,0 +1,382 @@
+instead using abstract theory on the properness of certain maps (phrased in terms of the filters
+`filter.cocompact`, `filter.cofinite`, etc) to deduce existence theorems, first to prove the
+existence of `g` maximizing `(g•z).im` (see `modular_group.exists_g_with_max_im`), and then among
+those, to minimize `|(g•z).re|` (see `modular_group.exists_g_with_given_cd_and_min_re`).
+-/
+
+open complex matrix matrix.special_linear_group upper_half_plane
+noncomputable theory
+
+local notation `SL(` n `, ` R `)`:= special_linear_group (fin n) R
+local prefix `↑ₘ`:1024 := @coe _ (matrix (fin 2) (fin 2) ℤ) _
+
+
+open_locale upper_half_plane complex_conjugate
+
+local attribute [instance] fintype.card_fin_even
+
+namespace modular_group
+
+section upper_half_plane_action
+
+/-- The action of `SL(2, ℤ)` on the upper half-plane, as a restriction of the `SL(2, ℝ)`-action. -/
+instance {R : Type*} [comm_ring R] [algebra R ℝ] : mul_action SL(2, R) ℍ :=
+mul_action.comp_hom ℍ (map (algebra_map R ℝ))
+
+lemma coe_smul (g : SL(2, ℤ)) (z : ℍ) : ↑(g • z) = num g z / denom g z := rfl
+lemma re_smul (g : SL(2, ℤ)) (z : ℍ) : (g • z).re = (num g z / denom g z).re := rfl
+@[simp] lemma smul_coe (g : SL(2, ℤ)) (z : ℍ) : (g : SL(2,ℝ)) • z = g • z := rfl
+
+@[simp] lemma neg_smul (g : SL(2, ℤ)) (z : ℍ) : -g • z = g • z :=
+show ↑(-g) • _ = _, by simp [neg_smul g z]
+
+lemma im_smul (g : SL(2, ℤ)) (z : ℍ) : (g • z).im = (num g z / denom g z).im := rfl
+
+lemma im_smul_eq_div_norm_sq (g : SL(2, ℤ)) (z : ℍ) :
+  (g • z).im = z.im / (complex.norm_sq (denom g z)) :=
+im_smul_eq_div_norm_sq g z
+
+@[simp] lemma denom_apply (g : SL(2, ℤ)) (z : ℍ) : denom g z = ↑ₘg 1 0 * z + ↑ₘg 1 1 := by simp
+
+end upper_half_plane_action
+
+section bottom_row
+
+/-- The two numbers `c`, `d` in the "bottom_row" of `g=[[*,*],[c,d]]` in `SL(2, ℤ)` are coprime. -/
+lemma bottom_row_coprime (g : SL(2, ℤ)) : is_coprime (↑ₘg 1 0) (↑ₘg 1 1) :=
+begin
+  use [- g 0 1, g 0 0],
+  have := det_fin_two g,
+  have := g.det_coe,
+  simp only [coe_fn_eq_coe] at *,
+  linarith
+end
+
+/-- Every pair `![c, d]` of coprime integers is the "bottom_row" of some element `g=[[*,*],[c,d]]`
+of `SL(2,ℤ)`. -/
+lemma bottom_row_surj :
+  set.surj_on (λ g : SL(2, ℤ), ↑ₘg 1) set.univ {cd | is_coprime (cd 0) (cd 1)} :=
+begin
+  rintros cd ⟨b₀, a, gcd_eqn⟩,
+  let A := ![![a, -b₀], cd],
+  have det_A_1 : det A = 1,
+  { convert gcd_eqn,
+    simp [A, det_fin_two, (by ring : a * (cd 1) + b₀ * (cd 0) = b₀ * (cd 0) + a * (cd 1))] },
+  refine ⟨⟨A, det_A_1⟩, set.mem_univ _, _⟩,
+  ext; simp [A]
+end
+
+end bottom_row
+
+section tendsto_lemmas
+
+open filter continuous_linear_map
+local attribute [instance] matrix.normed_group matrix.normed_space
+local attribute [simp] coe_smul
+
+/-- The function `(c,d) → |cz+d|^2` is proper, that is, preimages of bounded-above sets are finite.
+-/
+lemma tendsto_norm_sq_coprime_pair (z : ℍ) :
+  filter.tendsto (λ p : fin 2 → ℤ, ((p 0 : ℂ) * z + p 1).norm_sq)
+  cofinite at_top :=
```

Comment:
Can you split this proof into one about `ℝ` (assuming it's true!):
```suggestion
lemma tendsto_norm_sq_coprime_pair (z : ℍ) :
  filter.tendsto (λ p : fin 2 → ℝ, ((p 0 : ℂ) * z + p 1).norm_sq)
  cofinite at_top :=
```
and then a short proof that adapts this to be about `ℤ`?

Reply:
I prefer it the current way on stylistic grounds.  These are not lemmas of general interest, they are calculations best got out of the way as efficiently as possible.